feat: LightningGateway TTL

tvolk131 opened the pull request:
Fixes #2573

Open question: Should all DB record values be of type `LightningGatewayRegistration` even if they don't currently use the `valid_until`?

Edit: I talked with @m1sterc001guy offline and the short answer is yes.

m1sterc001guy commented on gateway/ln-gateway/src/state_machine/mod.rs:
```diff
@@ -471,35 +472,37 @@ impl GatewayClientModule {
-            gateway_id,
+    ) -> LightningGatewayRegistrationAnnouncement {
+        LightningGatewayRegistrationAnnouncement {
+            info: LightningGatewayInfo {
```
Nit: `LightningGatewayInfo` -> `LightningGateway`

tvolk131 replied:
Done.

m1sterc001guy commented on modules/fedimint-ln-common/src/lib.rs:
```diff
@@ -176,9 +176,63 @@ impl std::fmt::Display for LightningOutputOutcome {
+/// unsynchronized clocks which can each anchor the announcement to their local
+/// system time.
+#[derive(Debug, Clone, Serialize, Deserialize, Encodable, Decodable, PartialEq, Eq, Hash)]
+pub struct LightningGatewayRegistrationAnnouncement {
```
Nit: `LightningGatewayRegistrationAnnouncement` -> `LightningGatewayAnnouncement`

tvolk131 replied:
Done.

m1sterc001guy commented:
> Open question: Should all DB record values be of type LightningGatewayRegistration even if they don't currently use the valid_until?

I think we answered this offline today, we decided that they should (we might use the ttl in the future)

m1sterc001guy commented on modules/fedimint-ln-client/src/lib.rs:
```diff
@@ -214,12 +215,12 @@ impl LightningClientExt for Client {
                 .fetch_registered_gateways()
                 .await?
                 .into_iter()
-                .filter(|gw| !gw.is_expired())
```
Not filtering on `is_expired` anymore?

tvolk131 replied:
`fetch_registered_gateways()` now returns a list of `LightningGatewayAnnouncement`s which contain a `Duration`-based TTL, so checking for expired ones doesn't make sense since `Duration`s can't be negative.

m1sterc001guy commented on modules/fedimint-ln-common/src/lib.rs:
```diff
@@ -176,7 +176,61 @@ impl std::fmt::Display for LightningOutputOutcome {
+pub struct LightningGatewayRegistration {
+    pub info: LightningGateway,
+    /// Limits the validity of the announcement to allow updates
+    /// Anchored to local system time
```
Nit: Anchored does need to be caps

tvolk131 replied:
Done.

m1sterc001guy commented:
Overall looks good to me but I think some of the failing tests need debugging

tvolk131 commented:
> I think some of the failing tests need debugging

Fixed! The issue was that prior to this PR, `LightningGateway`s stored by each guardian were exactly the same since they each had identical `SystemTime`s for expiration, allowing a client to get all guardian's registered gateways and filter out duplicates by checking for exact equality. This invariant was broken by using `LightningGatewayAnnouncement` for sharing over the network, since the _unanchor -> transmit -> reanchor_ process allows for each guardian's re-anchored expiration time to be different. I added `filter_duplicate_gateways()` to address this.

m1sterc001guy commented on modules/fedimint-ln-common/src/api.rs:
```diff
@@ -127,3 +141,52 @@ where
+    // announcements for the same gateway ID, we only filter out announcements
+    // that have the same settings, keeping the one with the longest TTL.
+    gateways_by_gateway_id
+        .into_values()
```
A bit confused. Why do we create a hashmap above if we just throw away the keys and only use the values?

elsirion replied:
The gateways are still grouped by the keys (values are `Vec`s of GWs with the same id), even though they aren't used directly.

m1sterc001guy commented on modules/fedimint-ln-common/src/api.rs:
```diff
@@ -127,3 +141,52 @@ where
+                    .or_insert(ttl);
+            }
+
+            gateways.into_iter().map(|a| LightningGatewayAnnouncement {
```
Nit: use `announcement` instead of just `a`

tvolk131 replied:
Done.

elsirion commented on modules/fedimint-ln-common/src/api.rs:
```diff
@@ -127,3 +141,52 @@ where
+    gateways: Vec<LightningGatewayAnnouncement>,
+) -> Vec<LightningGatewayAnnouncement> {
+    let gateways_by_gateway_id =
+        gateways
+            .into_iter()
+            .fold(std::collections::HashMap::new(), |mut acc, gateway| {
+                acc.entry(gateway.info.gateway_id)
+                    .or_insert_with(Vec::new)
+                    .push(gateway);
+                acc
+            });
```
Could be replaced with [`Itertools::into_group_map`](https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.into_group_map).

tvolk131 replied:
Very cool! Done.

elsirion reviewed (approved):
This PR looks great! Very well documented and a good balance of functional and imperative style. Sorry that review took so long, too many conferences :see_no_evil:

elsirion commented:
Oh, needs a rebase :(

tvolk131 commented:
> Oh, needs a rebase :(

Done. This should be ready to submit!
